### PR TITLE
Import CSV

### PR DESCRIPTION
**Pull request by timifasubaa**
This PR allows users to import CSV files to superset. It extends the work that had already been done in https://github.com/apache/incubator-superset/pull/2381 and also adds the new functionality of uploading to a hive datasource.

### CODE REVIEW

**Comment by mistercrunch** [comment]
How does this relate to https://github.com/apache/incubator-superset/pull/3533 ?

**Comment by mistercrunch** [comment]
Build doesn't pass...

**Comment by timifasubaa** [comment]
The "Not a valid choice" error is coming up for postgres and mysql dbs (the 2 failing toxenvs). Still can't reproduce the error.

**Comment by timifasubaa** [comment]
@mistercrunch resolved!

**Comment by rpwils** [comment]
Will this be added to the next version?

**Comment by timifasubaa** [comment]
@mistercrunch any updates on this PR?

File: superset/db_engine_specs.py [hunk 1]
```diff
@@ -73,6 +82,66 @@ def extra_table_metadata(cls, database, table_name, schema_name):
...
+        df.to_sql(**kwargs)
+        table.user_id = g.user.id
+        table.schema = kwargs['schema']
+        table.fetch_metadata()
+        db.session.add(table)
+        db.session.commit()
+
+    @staticmethod
+    def create_table_from_csv(form, table):
+        def allowed_file(filename):
```
Comment (john-bodley) [inline superset/db_engine_specs.py]: Nit. Rename `allowed_file` to _allowed_file`.

File: superset/db_engine_specs.py [hunk 2]
```diff
@@ -717,6 +786,51 @@ def fetch_result_sets(cls, db, datasource_type, force=False):
...
+            upload_path, 'airbnb-superset',
+            os.path.join(upload_prefix, table_name, filename))
+        sql = """CREATE EXTERNAL TABLE {table_name} ( {schema_definition} )
+            ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' STORED AS
+            TEXTFILE LOCATION '{location}'""".format(**locals())
+        try:
+            logging.info(form.con.data)
+            engine = create_engine(form.con.data)
+            engine.execute(sql)
+            return (True, '')
```
Comment (john-bodley) [inline superset/db_engine_specs.py]: It seems more Pythonic to return a scalar and raise an exception in the case of the failure.

File: superset/views/core.py [hunk 3]
```diff
@@ -314,6 +318,65 @@ class DatabaseAsync(DatabaseView):
...
+        form.mangle_dupe_cols.data = True
+        form.skipinitialspace.data = False
+        form.skip_blank_lines.data = True
+        form.parse_dates.data = True
+        form.infer_datetime_format.data = True
+        form.decimal.data = '.'
+        form.if_exists.data = 'append'
+        all_datasources = db.session.query(
+            models.Database.sqlalchemy_uri,
+            models.Database.database_name)\
```
Comment (john-bodley) [inline superset/views/core.py]: Nit. I don't think the `\` is needed.

File: superset/views/core.py [hunk 4]
```diff
@@ -314,6 +318,65 @@ class DatabaseAsync(DatabaseView):
...
+            .all()
+        form.con.choices += all_datasources
+
+    def form_post(self, form):
+        def upload_file(csv_file):
```
Comment (john-bodley) [inline superset/views/core.py]: Nit. Rename `upload_file` to `_upload_file`.

File: tests/core_tests.py [hunk 5]
```diff
@@ -786,6 +788,46 @@ def test_viz_get_fillna_for_columns(self):
             {'name': ' NULL', 'sum__num': 0},
         )
 
+    def test_import_csv(self):
```
Comment (john-bodley) [inline tests/core_tests.py]: Thanks for adding unit tests!

File: tox.ini [hunk 6]
```diff
@@ -68,12 +68,12 @@ commands =
 [testenv:py27-mysql]
 basepython = python2.7
 setenv =
-    SUPERSET__SQLALCHEMY_DATABASE_URI = mysql://mysqluser:mysqluserpassword@localhost/superset?charset=utf8
+    SUPERSET__SQLALCHEMY_DATABASE_URI = mysql://root@localhost/superset?charset=utf8
```
Comment (john-bodley) [inline tox.ini]: Why are we changing the user/password for the test databases?

Comment (timifasubaa) [reply tox.ini]: It didn't work with the previous credentials. I think insufficient privileges for adding a table if I recall correctly.

File: superset/views/core.py (hunk 3 above)
Comment (timifasubaa) [reply superset/views/core.py]: When I removed it, I got unexpected indent.

**Comment by john-bodley** [comment]
LGTM

**Comment by TryHarder01** [comment]
dumb question... is there an obvious way to see what version this PR became available?

**Comment by frankhsu** [comment]
@TryHarder01 Search for the PR number #3643 in the CHANGELOG.

**Comment by TryHarder01** [comment]
@frankhsu no results when I search the change log. 

Not a huge deal, I was just curious how to track PRs into versions.

**Comment by mistercrunch** [comment]
A release is coming soon :)

**Pull request merged**